Improve <textarea> validation performance

darinadler opened the pull request:
#### 95861845546f13626a0ea23d6eab4e513ff6f760
<pre>
Improve &lt;textarea&gt; validation performance
<a href="https://bugs.webkit.org/show_bug.cgi?id=248447">https://bugs.webkit.org/show_bug.cgi?id=248447</a>
rdar://problem/102909909

Reviewed by Ryosuke Niwa.

Added laziness to the HTMLTextAreaElement computation of validity. Also did style
tweaks and minor reorganization of the two text form element classes.

* Source/WebCore/html/FileInputType.cpp: Added include of FileChooser.h.
(WebCore::FileInputType::applyFileChooserSettings): Update to pass reference.

* Source/WebCore/html/FormController.cpp: Added include of FileChooser.h.

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::capsLockIndicatorElement const): Deleted.
(WebCore::HTMLInputElement::accept const): Deleted.
(WebCore::HTMLInputElement::isInRequiredRadioButtonGroup): Deleted.
(WebCore::HTMLInputElement::setRangeText): Removed unneeded overload and let
the base class version of setRangeText call the virtual one.

* Source/WebCore/html/HTMLInputElement.h: Removed unneeded includes.
Moved the functions for the IDL file to the top of the class, sorted in the
order that matches the IDL. Removed a few unused functions. Made three
public functions private. Tweaked comments.

* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::HTMLTextAreaElement): Removed the tag name argument.
(WebCore::HTMLTextAreaElement::create): Updated to not pass the tag name to the constructor.
(WebCore::HTMLTextAreaElement::formControlType const): Use textareaTag so we don&apos;t need a global.
(WebCore::HTMLTextAreaElement::collectPresentationalHintsForAttribute): Moved body of the
shouldWrapText in.
(WebCore::HTMLTextAreaElement::parseAttribute): Moved the bodies of maxLengthAttributeChanged,
and minLengthAttributeChanged in.
(WebCore::HTMLTextAreaElement::maxLengthAttributeChanged): Deleted.
(WebCore::HTMLTextAreaElement::minLengthAttributeChanged): Deleted.
(WebCore::HTMLTextAreaElement::appendFormData): Tweaked coding style.
(WebCore::HTMLTextAreaElement::hasCustomFocusLogic const): Moved to the class definition.
(WebCore::HTMLTextAreaElement::defaultTabIndex const): Ditto.
(WebCore::HTMLTextAreaElement::isKeyboardFocusable const): Ditto.
(WebCore::HTMLTextAreaElement::isMouseFocusable const): Ditto.
(WebCore::HTMLTextAreaElement::handleBeforeTextInsertedEvent const): Call maxLength. Merge in the
sanitizeUserInputValue function.
(WebCore::HTMLTextAreaElement::sanitizeUserInputValue): Deleted.
(WebCore::HTMLTextAreaElement::rendererWillBeDestroyed): Moved to the class definition.
(WebCore::HTMLTextAreaElement::validationMessage const): Call maxLength.
(WebCore::HTMLTextAreaElement::valueMissing const): Pass null string for value, so the value will
be computed lazily. In the overload that takes a StringView, get the value lazily.
(WebCore::HTMLTextAreaElement::tooShort const): Ditto.
(WebCore::HTMLTextAreaElement::tooLong const): Ditto.
(WebCore::HTMLTextAreaElement::isValidValue const): Change type to StringView.
(WebCore::HTMLTextAreaElement::shouldUseInputMethod): Moved to the class definition.
(WebCore::HTMLTextAreaElement::placeholderElement const): Ditto.
(WebCore::HTMLTextAreaElement::matchesReadWritePseudoClass const): Ditto.
(WebCore::HTMLTextAreaElement::willRespondToMouseClickEventsWithEditability const): Ditto.

* Source/WebCore/html/HTMLTextAreaElement.h: Removed unneeded include. Moved the functions for the
IDL file to the top of the class, sorted in the order that matches the IDL. Removed some functions.
Moved some function bodies here. Use StringView more often.

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::setRangeText): Take StringView.

* Source/WebCore/html/HTMLTextFormControlElement.h: Changed setRangeText to take a StringView.
Also made the single argument form non-virtual.

* Source/WebCore/html/InputType.h:
(WebCore::InputType::capsLockIndicatorElement const): Deleted.
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::capsLockIndicatorElement const): Deleted.
* Source/WebCore/html/TextFieldInputType.h: Deleted capsLockIndicatorElement.

* Source/WebCore/loader/EmptyClients.cpp: Deleted unneeded includes.
* Source/WebCore/page/Chrome.cpp: Ditto.

* Source/WebCore/platform/FileChooser.cpp:
(WebCore::FileChooser::FileChooser): Take a reference.
(WebCore::FileChooser::create): Ditto.

* Source/WebCore/platform/FileChooser.h: Updated for the above.

* Source/WebKitLegacy/mac/DOM/DOMHTMLInputElement.mm:
(-[DOMHTMLInputElement setRangeText:]): Create a String explicitly since the function now
takes a StringView.
(-[DOMHTMLInputElement setRangeText:start:end:selectionMode:]): Ditto.
* Source/WebKitLegacy/mac/DOM/DOMHTMLTextAreaElement.mm:
(-[DOMHTMLTextAreaElement setRangeText:]): Ditto.
(-[DOMHTMLTextAreaElement setRangeText:start:end:selectionMode:]): Ditto.

Canonical link: <a href="https://commits.webkit.org/257417@main">https://commits.webkit.org/257417@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/450484e953a93cdd1a312fe88c0b9ee1a20ec53d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98949 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8157 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32111 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108362 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168615 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102907 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8710 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91480 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106332 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104650 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33622 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88417 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76471 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2062 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1968 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5103 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6928 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/42514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3376 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/HTMLTextAreaElement.h:
```diff
@@ -119,28 +96,46 @@ class HTMLTextAreaElement final : public HTMLTextFormControlElement {
     void updateFocusAppearance(SelectionRestorationMode, SelectionRevealMode) final;
 
-    bool accessKeyAction(bool sendMouseEvents) final;
+    bool accessKeyAction(bool) final { focus(); return false; }
```
I think it would be better if `focus()` and `return false` were on their own lines.

rniwa commented on Source/WebCore/html/HTMLTextAreaElement.h:
```diff
@@ -119,28 +96,46 @@ class HTMLTextAreaElement final : public HTMLTextFormControlElement {
     RefPtr<HTMLElement> m_placeholder;
     mutable String m_value;
+    WrapMethod m_wrap { SoftWrap };
     mutable bool m_isDirty { false };
```
We should probably change these to uint8_t so that they're tightly packed on Windows.

darinadler committed:
Improve <textarea> validation performance
https://bugs.webkit.org/show_bug.cgi?id=248447
rdar://problem/102909909

Reviewed by Ryosuke Niwa.

Added laziness to the HTMLTextAreaElement computation of validity. Also did style
tweaks and minor reorganization of the two text form element classes.

* Source/WebCore/html/FileInputType.cpp: Added include of FileChooser.h.
(WebCore::FileInputType::applyFileChooserSettings): Update to pass reference.

* Source/WebCore/html/FormController.cpp: Added include of FileChooser.h.

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::capsLockIndicatorElement const): Deleted.
(WebCore::HTMLInputElement::accept const): Deleted.
(WebCore::HTMLInputElement::isInRequiredRadioButtonGroup): Deleted.
(WebCore::HTMLInputElement::setRangeText): Removed unneeded overload and let
the base class version of setRangeText call the virtual one.

* Source/WebCore/html/HTMLInputElement.h: Removed unneeded includes.
Moved the functions for the IDL file to the top of the class, sorted in the
order that matches the IDL. Removed a few unused functions. Made three
public functions private. Tweaked comments.

* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::HTMLTextAreaElement): Removed the tag name argument.
(WebCore::HTMLTextAreaElement::create): Updated to not pass the tag name to the constructor.
(WebCore::HTMLTextAreaElement::formControlType const): Use textareaTag so we don't need a global.
(WebCore::HTMLTextAreaElement::collectPresentationalHintsForAttribute): Moved body of the
shouldWrapText in.
(WebCore::HTMLTextAreaElement::parseAttribute): Moved the bodies of maxLengthAttributeChanged,
and minLengthAttributeChanged in.
(WebCore::HTMLTextAreaElement::maxLengthAttributeChanged): Deleted.
(WebCore::HTMLTextAreaElement::minLengthAttributeChanged): Deleted.
(WebCore::HTMLTextAreaElement::appendFormData): Tweaked coding style.
(WebCore::HTMLTextAreaElement::hasCustomFocusLogic const): Moved to the class definition.
(WebCore::HTMLTextAreaElement::defaultTabIndex const): Ditto.
(WebCore::HTMLTextAreaElement::isKeyboardFocusable const): Ditto.
(WebCore::HTMLTextAreaElement::isMouseFocusable const): Ditto.
(WebCore::HTMLTextAreaElement::handleBeforeTextInsertedEvent const): Call maxLength. Merge in the
sanitizeUserInputValue function.
(WebCore::HTMLTextAreaElement::sanitizeUserInputValue): Deleted.
(WebCore::HTMLTextAreaElement::rendererWillBeDestroyed): Moved to the class definition.
(WebCore::HTMLTextAreaElement::validationMessage const): Call maxLength.
(WebCore::HTMLTextAreaElement::valueMissing const): Pass null string for value, so the value will
be computed lazily. In the overload that takes a StringView, get the value lazily.
(WebCore::HTMLTextAreaElement::tooShort const): Ditto.
(WebCore::HTMLTextAreaElement::tooLong const): Ditto.
(WebCore::HTMLTextAreaElement::isValidValue const): Change type to StringView.
(WebCore::HTMLTextAreaElement::shouldUseInputMethod): Moved to the class definition.
(WebCore::HTMLTextAreaElement::placeholderElement const): Ditto.
(WebCore::HTMLTextAreaElement::matchesReadWritePseudoClass const): Ditto.
(WebCore::HTMLTextAreaElement::willRespondToMouseClickEventsWithEditability const): Ditto.

* Source/WebCore/html/HTMLTextAreaElement.h: Removed unneeded include. Moved the functions for the
IDL file to the top of the class, sorted in the order that matches the IDL. Removed some functions.
Moved some function bodies here. Use StringView more often.

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::setRangeText): Take StringView.

* Source/WebCore/html/HTMLTextFormControlElement.h: Changed setRangeText to take a StringView.
Also made the single argument form non-virtual.

* Source/WebCore/html/InputType.h:
(WebCore::InputType::capsLockIndicatorElement const): Deleted.
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::capsLockIndicatorElement const): Deleted.
* Source/WebCore/html/TextFieldInputType.h: Deleted capsLockIndicatorElement.

* Source/WebCore/loader/EmptyClients.cpp: Deleted unneeded includes.
* Source/WebCore/page/Chrome.cpp: Ditto.

* Source/WebCore/platform/FileChooser.cpp:
(WebCore::FileChooser::FileChooser): Take a reference.
(WebCore::FileChooser::create): Ditto.

* Source/WebCore/platform/FileChooser.h: Updated for the above.

* Source/WebKitLegacy/mac/DOM/DOMHTMLInputElement.mm:
(-[DOMHTMLInputElement setRangeText:]): Create a String explicitly since the function now
takes a StringView.
(-[DOMHTMLInputElement setRangeText:start:end:selectionMode:]): Ditto.
* Source/WebKitLegacy/mac/DOM/DOMHTMLTextAreaElement.mm:
(-[DOMHTMLTextAreaElement setRangeText:]): Ditto.
(-[DOMHTMLTextAreaElement setRangeText:start:end:selectionMode:]): Ditto.

Canonical link: https://commits.webkit.org/257417@main
